### ISSUE TITLE
EF7 is not saving navigation object on database

### ISSUE THREAD

**Issue by Bartmax** [dotnet/efcore#2387]
I'm not sure if I'm doing something wrong or is not supported yet. I was trying to figure out what's going on with no clue.

the problem: In the database, the AppointmentRequest is saved but AppointmentRequestLine is empty.

this is the code:

```
public class ApplicationDbContext : IdentityDbContext<ApplicationUser, IdentityRole<Guid>, Guid>
{
    public DbSet<AppointmentRequest> AppointmentRequests { get; set; }
}

public class AppointmentRequest
{
    public Guid AppointmentRequestId { get; private set; }
    public ICollection<AppointmentRequestLine> Lines { get; private set; }
}

public class AppointmentRequestLine
{
    public AppointmentRequestLine(string label, string value){
        Label = label;
        Value = value;
    }
    public string Label { get; private set; }
    public string Value { get; private set; }
    public int AppointmentRequestLineId { get; private set; }
}

public async Task SaveAsync(AppointmentRequest request)
{
    context.AppointmentRequests.Add(request);
    await context.SaveChangesAsync();
}

and calling it for testing as:

var lines = new List<AppointmentRequestLine>();
var line = new AppointmentRequestLine("hello", "world");
lines.Add(line);

var req = new AppointmentRequest(lines);

await appointmentService.SaveAsync(req);
```

The database is created using migrations (had to delete everything and create 1 migration because a second one fails but that another issue)

Is there anything explicit I should do to save the complete graph? Debugging shows that the lines has 1 item in memory before the SaveChanges().

No error at all, it just saves the AppointmentRequest but no AppointmentRequestLine.

**Comment by Bartmax** [comment]
This new method for saving solved the issue. Is this supposed to be explicit as this or is because is not yet implemented ?

```
public async Task SaveAsync(AppointmentRequest request)
{
    context.AppointmentRequests.Add(request);
    foreach (var item in request.Lines)
    {
        context.Entry(item).State = item.AppointmentRequestLineId == default(int) ? EntityState.Added : EntityState.Modified;
    }
    await context.SaveChangesAsync();
}
```

**Comment by MatthewMWR** [comment]
My understanding is that the default in EF7 is to attach the single explicit item. To attach the object graph as in old versions you can do manually like you have done here or use AttachGraph()

**Comment by rowanmiller** [comment]
Correct. Generally in .NET APIs `XyzSet.Add` only adds the root and the fact past versions of EF added the entire graph has been a source of confusion for folks. That said, given that's what `Add` did in past releases we are discussing renaming the API to something more like `AddSingleObject` (hopefully we come up with a better name) to avoid confusion.

As @MatthewMWR mentioned there is also the AttachGraph method to make things a little simpler.

**Comment by Bartmax** [comment]
Okay thank you. I will take a look at attach graph. 

My confusion was because I expected to work like previous versions and the lack of documentation. This are beta bits so it's normal to have this issues at this stage.

While I agree the api may use a new name, making it explicitly clear on documentation on RTM may be enough. It's something one will hit only once.
Another idea is to have 2 methods (to avoid confusion) like `AddSingle` and `AddGraph` ? (if you ever want/is-worth to create an `AddGraph` method, of course)

I'm confident you will find a great name/solution.

I personally find this 'new way' a great decision.

**Comment by JohnyL** [comment]
@rowanmiller
I find the EF6's behavior is more natural because I explicitly connect related entity to parent entity. Once again - this is something that is done __intentionally__.
So, for instance we have two entities - Blog and Post. The relationship is mandatory (one-to-one-or-many), i.e. Post can't exist without Blog. Say, I have the following code:
```C#
var post1 = new Post { Id = 1, Title = "Title1", BlogId = 1 };
var post2 = new Post { Id = 2, Title = "Title2", BlogId = 1 };
var blog = new Blog { Id = 1, Title = "Blog Title" };
// Attach new posts to context
context.Posts.Add(post1);
context.Posts.Add(post2);
// Attach new posts to blog
blog.Posts.Add(post1);
blog.Posts.Add(post2);
// Attach blog to context
context.Blogs.Add(blog);
context.SaveChanges();
```
What happens under the hood when saving blog? I guess, EF can't insert posts when there's no related blog. Blog must be inserted first, and then - blog. So, what's the point in explicitly attaching posts to context when EF anyway scans for relationships?

**Comment by rowanmiller** [comment]
@JohnyL we're looking at enabling both and having something like `AddSingleObject` and `AddGraph` rather than just `Add` so that folks can make a conscious choice about which one they want.

**Comment by JohnyL** [comment]
@rowanmiller That would be great! Thanks a lot! :)

**Comment by Agalla81** [comment]
Not sure if it's still under construction, but maybe "AddRoot" vs "AddGraph"?
